### PR TITLE
OSASINFRA-3535: Add support for Hypershift to Cinder CSI

### PR DESCRIPTION
**Pull request by stephenfin**
In #278, #275 and #298, we worked to integrate Cinder CSI driver support here, into `csi-operator`. With that out of the way, we can now move onto working on Hypershift support.

The changes required here are mostly minimal: we need to ensure that we sync our config map contain configuration for the drivers to the correct namespace. We also need to ensure that we pass through the guest kubeconfig to the controllers. All of this has been done elsewhere and should not be surprising. The main change is to our syncer. I would suggest reading the commit messages for more context, but the tl;dr: is that we modify this so that (a) it's able to run without a source config map, and (b) it's able to dump the results in multiple locations.

**Related changes**

- https://github.com/openshift/cluster-storage-operator/pull/525
- https://github.com/openshift/hypershift/pull/4936

### CODE REVIEW

**Review by MaysaMacedo — CHANGES_REQUESTED** [review]
The following check also needs to be removed to support Hypershift
https://github.com/openshift/csi-operator/pull/302/files#diff-9e4d5abc381c99fd5db77cd5b530830b31ceba8d952f57da3f095fa3bd7fcac8R120-R123

**Comment by EmilienM** [comment]
/testwith openshift/hypershift/main/e2e-openstack-csi-cinder openshift/csi-operator#302 openshift/cluster-storage-operator#525

File: pkg/openstack-cinder/config/configsync.go
Comment (MaysaMacedo) [inline pkg/openstack-cinder/config/configsync.go:97]: is it really needed in the control plane namespace?
I took a look on the assets generated and the DaemonSet that runs on the guest cluster seems to be the only one that needs it https://github.com/openshift/csi-operator/pull/302/files#diff-dece87039665ecf71f9003b48e52ce76685b59171f7c8e74ae0110f72b0526acR197.

Comment (stephenfin) [reply pkg/openstack-cinder/config/configsync.go:97]: The diff for `controller.yaml` is hidden by default so you need to expand that. When you do, you'll see it's used in the controller also.

https://github.com/openshift/csi-operator/pull/302/files#diff-8114e4393b7e73a8d7b1a309b1d0688968e54ffe7b996087cd410b780d386731R412

**Comment by stephenfin** [comment]
/retest-required

/hold

I have tested this locally in a HCP context and the deployment succeeded. I'm not sure if I missed something in the standalone context so I'm deploying that now locally also. Let's recheck in the CI also to see if it was just a timeout.

**Comment by stephenfin** [comment]
/unhold

Figured the issue out and reshuffled things around so that I was retrieving the `infrastructure` CR inside the `sync` loop. This should be good to go now.

File: pkg/openstack-cinder/config/configsync.go
Comment (mandre) [inline pkg/openstack-cinder/config/configsync.go:82]: Shouldn't it be `c.KubeClient` ?

**Review by mandre — COMMENTED** [review]
I'll need to understand better how the new configmap syncer works in practice and look at CI results more closely. Can it cause issues on upgraded clusters?

File: pkg/openstack-cinder/config/configsync.go
Comment (stephenfin) [reply pkg/openstack-cinder/config/configsync.go]: It should, yes. I reworked this yesterday and must have fat-fingered this.

**Comment by stephenfin** [comment]
> I'll need to understand better how the new configmap syncer works in practice and look at CI results more closely. Can it cause issues on upgraded clusters?

I called this out in the commit message, but the largest change in the config syncer is that we now only copy the `[BlockStorage]` section, meaning we ignore/drop the `[Global]`, `[LoadBalancer]`, and `[Metadata]` sections. I think this is definitely the correct thing to do for `[Global]` and `[LoadBalancer]`, since we don't want users providing auth info outside of the `clouds.yaml` file managed by cloud-credential-operator (among other things, doing so would make credential rotation more difficult) and load balancer configuration is irrelevant for the CSI drivers. The removal of support for configuring the `[Metadata]` section is a little more nuanced but again, I don't know if we want people tweaking this from [its default of `configDrive,metadataService`](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md#metadata).

The only way I can envision this causing issue on an upgraded cluster is if either (a) users had *completely* overridden the auth info found in `clouds.yaml` with other credentials or (b) they had *partially* overridden the auth info found in `clouds.yaml`. Both seem unlikely. In the former case, the new user would need to be an admin or a set of application credentials for the same user or they wouldn't have permissions to attach/detach volumes from the instance. In the latter case, the user would also have needed to do this for CCM and every other operator or service that talks to OpenStack. As such, I don't think we need to worry about this.

That change aside, the other change is that we have obviously reworked things to support HCP. The crux of that is that while in standalone we source from one of two potential config maps and save the results to a config map in the operator namespace (`openshift-cluster-csi-drivers / cloud-conf`), in HCP we don't source from anywhere (by design) and save the results to *two* config maps, one in the management cluster (`clusters-${NAME} / cloud-conf`) and one in the guest cluster (`openshift-cluster-csi-drivers / cloud-conf`). This is done because both the controllers (running in the management cluster) and the drivers (running in the guest cluster) need a configuration file (even though I don't think the former actually uses the thing).

Let me know if any of that isn't clear

**Comment by stephenfin** [comment]
/retest

At least `ci/prow/e2e-openstack` failed due to infra flake.

**Comment by mandre** [comment]
@stephenfin we'll need to update openstack-test to match us dropping the global section of the cloud.conf: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_csi-operator/302/pull-ci-openshift-csi-operator-master-e2e-openstack/1857356418852589568

**Comment by EmilienM** [comment]
/testwith openshift/hypershift/main/e2e-openstack-csi-cinder openshift/hypershift#4936

**Comment by stephenfin** [comment]
/hold

While we wait on https://github.com/openshift/openstack-test/pull/202 to land and get a green CI :heavy_check_mark: for OpenStack jobs.

The other two failures are long-term failures that we need to address separately.

**Comment by EmilienM** [comment]
/testwith openshift/hypershift/main/e2e-openstack-csi-cinder https://github.com/openshift/hypershift/pull/4936

**Comment by smg247** [comment]
/testwith openshift/hypershift/main/e2e-openstack-csi-cinder https://github.com/openshift/hypershift/pull/4936

**Comment by EmilienM** [comment]
@stephenfin This job was running in Mecha where a self signed certificate is being used and look at these logs (this looks like the issue I had in my environment, as well):

https://github.com/openshift/csi-operator/pull/302/checks?check_run_id=33062602418

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/multi-pr-openshift-csi-operator-302-openshift-hypershift-4936-e2e-openstack-csi-cinder/1857514674782736384/artifacts/e2e-openstack-csi-cinder/dump/artifacts/namespaces/clusters-9337ef3317244feda749/core/pods/logs/openstack-cinder-csi-driver-operator-687c8496d5-nzdzm-openstack-cinder-csi-driver-operator.log

**Comment by stephenfin** [comment]
/testwith openshift/hypershift/main/e2e-openstack-csi-cinder https://github.com/openshift/hypershift/pull/4936

**Comment by EmilienM** [comment]
Again: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/multi-pr-openshift-csi-operator-302-openshift-hypershift-4936-e2e-openstack-csi-cinder/1857734832038088704/artifacts/e2e-openstack-csi-cinder/dump/artifacts/namespaces/clusters-28ff1dd1f48d163bdbf2/core/pods/logs/openstack-cinder-csi-driver-operator-f7998dc98-wp9wt-openstack-cinder-csi-driver-operator.log

**Comment by stephenfin** [comment]
/hold remove

https://github.com/openshift/openstack-test/pull/202 has landed.

**Comment by stephenfin** [comment]
> Again: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/multi-pr-openshift-csi-operator-302-openshift-hypershift-4936-e2e-openstack-csi-cinder/1857734832038088704/artifacts/e2e-openstack-csi-cinder/dump/artifacts/namespaces/clusters-28ff1dd1f48d163bdbf2/core/pods/logs/openstack-cinder-csi-driver-operator-f7998dc98-wp9wt-openstack-cinder-csi-driver-operator.log

Indeed. https://github.com/openshift/cluster-storage-operator/pull/538 should fix that.

File: pkg/openstack-cinder/config/cloudinfo.go
Comment (MaysaMacedo) [inline pkg/openstack-cinder/config/cloudinfo.go:88]: it's worth creating a constant for the path `/etc/kubernetes/static-pod-resources/configmaps/cloud-config`

Comment (stephenfin) [reply pkg/openstack-cinder/config/cloudinfo.go:88]: I'll move this if I need to respin.

File: pkg/openstack-cinder/config/cloudinfo.go
Comment (mandre) [inline pkg/openstack-cinder/config/cloudinfo.go:91]: Isn't `/etc/kubernetes/static-pod-resources/configmaps/cloud-config` a directory? If so we'll need to point to the file holding the ca-cert...

**Comment by stephenfin** [comment]
/testwith openshift/hypershift/main/e2e-openstack-csi-cinder https://github.com/openshift/hypershift/pull/4936

**Comment by EmilienM** [comment]
/testwith openshift/hypershift/main/e2e-openstack openshift/hypershift#4936
testing this one too as it's broken and we need it back sooner rather than later

**Comment by EmilienM** [comment]
all changes here are only for OpenStack Cinder except a minor typo fix.

/approve

**Comment by EmilienM** [comment]
/testwith openshift/hypershift/main/e2e-openstack https://github.com/openshift/hypershift/pull/4936

**Comment by EmilienM** [comment]
/testwith openshift/hypershift/main/e2e-openstack-csi-cinder https://github.com/openshift/hypershift/pull/4936
infra error

**Comment by EmilienM** [comment]
/testwith openshift/hypershift/main/e2e-openstack openshift/hypershift#4936

**Comment by EmilienM** [comment]
let's merge what we have now and follow-up with bug fixes that are found in the HCP CI jobs.
/hold cancel

**Comment by EmilienM** [comment]
> /retest-required

I think we're good. Waiting on QE approval and this is good to be shipped. We'll have PRs on top as bug fixes.

**Comment by mandre** [comment]
> I think we're good. Waiting on QE approval and this is good to be shipped. We'll have PRs on top as bug fixes.

@rlobillo whenever you get a chance.

**Comment by rlobillo** [comment]
/label qe-approved

e2e-openstack failure is unrelated

**Pull request merged**
